### PR TITLE
deps: update the sdk 50 branch to main

### PR DESCRIPTION
**Pull request by faddat**
## Description

This PR merges the sdk 50 branch to main.  It is comprehensive, and a secondary PR will be created to update only the callbacks module, as well as another to update cosmos-sdk only. 

### Commit Message / Changelog Entry

```bash
deps: update sdk 50 branch to main
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

## includes

- build(deps): Bump bufbuild/buf-setup-action from 1.25.0 to 1.25.1 (#4286)
- deps: bump golangci-lint to v1.53.3 (#4276)
- testing: fix usage on TimeoutPacket to use counterparty portID/channelID in nextSeqRecv query (#4319)
- build(deps): Bump bufbuild/buf-setup-action from 1.25.1 to 1.26.0 (#4321)
- imp: use `types.MetadataFromVersion` helper function for callback handlers (#4290)
- feat(callbacks): adr8 implementation (#3939)
- Duplicate SimApp into callbacks directory (#4337)
- Remove callbacks wiring in ibc go simapp (#4340)
- Give callbacks its own go.mod (#4341)
- imp(callbacks): remove reconstructed packet logic from 'SendPacket' (#4343)
- fix(proto): fix nested msg signer annoation (#4336)
- switched back to cosmos/relayer (#4345)
- build(deps): Bump golangci/golangci-lint-action from 3.6.0 to 3.7.0 (#4349)
- build(deps): Bump bufbuild/buf-setup-action from 1.26.0 to 1.26.1 (#4350)
- chore: use IsOpen method in missed places (#4354)
- end of life for v4.1, v4.2, v4.3 (#4335)
- Add linting for callbacks submodule. (#4355)
- docs(simapp): fixed inline middleware wiring comments (#4361)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

**Comment by faddat** [comment]
I think that for now we should try to exclude the callbacks module. How tightly wired in is it to other components of IBC?

**Review by crodriguezvega — APPROVED** [review]
This looks good to me. Thank you, @faddat!

**Comment by faddat** [comment]
oh interesting -- I did not realize that this could be considered g2g -- I will have a poke tomorrow.  Thanks @crodriguezvega !

**Comment by faddat** [comment]
also if we want to make the linter truly happy and easy to use then we'd add a long-running go.work file and keep go.work.sum in .gitignore.  Osmosis does this and I love it.

**Comment by colin-axner** [comment]
I fixed some of the callback issues updating to SDK v0.50. Didn't get time to fix the rest of the updates required in the test files, but should hopefully be straight forward

**Comment by faddat** [comment]
I'll have a look, thank you!

**Comment by faddat** [comment]
We're good to go here :)

File: .github/workflows/golangci.yml [hunk 1]
```diff
@@ -20,7 +20,11 @@ jobs:
           go-version: '1.21'
       - uses: actions/checkout@v3
       - name: golangci-lint
-        uses: golangci/golangci-lint-action@v3.6.0
+        uses: golangci/golangci-lint-action@v3.7.0
         with:
           version: v1.54.2
+<<<<<<< HEAD
+          args: --timeout 5m
+=======
```
Comment (colin-axner) [inline .github/workflows/golangci.yml:28]: needs fix

**Comment by faddat** [comment]
hey thank you @colin-axner -- I committed, thought "hey its done" and stepped away for a call :).

**Comment by colin-axner** [comment]
timing worked out well :) I had the other commits ready to go!

**Review by colin-axner — APPROVED** [review]
LGTM! Pending passing tests for everything except e2e's. Reviewed mostly everything, I expect any unnecessary changes to be caught in review of merging feature branch to main

**Comment by colin-axner** [comment]
merging! Thanks everyone!

**Pull request merged**